### PR TITLE
ARROW-107: [C++] Implement IPC for structs

### PR DESCRIPTION
**Pull request by emkornfield**
Some other changes (I tried to isolate each in there own commit):
1.  Changed NumericTypes to be its own tempated type instead of separate macros (this made debugging easier)
2.  Fix an existing unit test for IPC that row counts inconsistent with row batch size.
3.  Some minor make-format changes.

### CODE REVIEW

**Comment by emkornfield** [comment]
it looks like the test failed due to the the flaky JVM test ..

File: cpp/src/arrow/types/primitive.h [hunk 1]
```diff
@@ -53,43 +53,42 @@ class ARROW_EXPORT PrimitiveArray : public Array {
...
-    }                                                                                  \
-                                                                                       \
-    const T* raw_data() const { return reinterpret_cast<const T*>(raw_data_); }        \
-                                                                                       \
-    T Value(int i) const { return raw_data()[i]; }                                     \
-  };
+template <class TypeClass, class T>
+class ARROW_EXPORT NumericArray : public PrimitiveArray {
+ public:
+  using value_type = T;
```
Comment (wesm) [inline cpp/src/arrow/types/primitive.h]: We may be able to further simplify by using `TypeClass::c_type` here and omitting `class T` from the template args

**Comment by wesm** [comment]
lgtm outside the template nit -- thank you for doing that refactoring.

**Comment by wesm** [comment]
In theory the build should pass now that ARROW-260 is merged

File: cpp/src/arrow/types/primitive.h (hunk 1 above)
Comment (emkornfield) [inline cpp/src/arrow/types/primitive.h]: Good idea.  pushing a new commit in a few minutes.

**Comment by wesm** [comment]
+1, thank you

**Pull request closed without merging**